src/modals/cluster-upgrade/TrackProgress.js

```diff
@@ -0,0 +1,119 @@
+import React, { useState, useEffect } from 'react'
+import PropTypes from 'prop-types'
+import { msg } from '_/intl-messages'
+import { trackUpgradeProgress, cancelTracker } from './data'
+
+import {
+  Bullseye,
+  Button,
+  EmptyState,
+  EmptyStateBody,
+  EmptyStateIcon,
+  EmptyStateSecondaryActions,
+  Progress,
+  Title,
+} from '@patternfly/react-core'
+import { CogsIcon } from '@patternfly/react-icons'
+// import { LogViewer } from '@patternfly/react-log-viewer'
+
+const STATUS_TO_TITLE = {
+  'pending': (clusterName) => msg.clusterUpgradeOperationPending({ clusterName }),
+  'started': (clusterName) => msg.clusterUpgradeOperationStarted({ clusterName }),
+  'complete': (clusterName) => msg.clusterUpgradeOperationComplete({ clusterName }),
+  'failed': (clusterName) => msg.clusterUpgradeOperationFailed({ clusterName }),
+}
+
+const STATUS_TO_PROGRESS_VARIANT = {
+  'pending': null,
+  'started': null,
+  'complete': 'success',
+  'failed': 'warning',
+}
+
+const TrackProgress = ({
+  cluster,
+  correlationId,
+  onJumpToEvents,
+  onClose,
+}) => {
+  const [status, setStatus] = useState('pending')
+  const [percent, setPercent] = useState(0)
+  const [log, setLog] = useState([])
+
+  // track status once a correlation id is set (or changed)
+  useEffect(() => {
+    if (!correlationId) {
+      return
+    }
+
+    setStatus('started')
+    trackUpgradeProgress(correlationId, ({ isRunning, percent, log }) => {
+      setStatus(isRunning ? 'started' : 'complete')
+      setPercent(percent)
+      setLog(currentLog => [...currentLog, ...log])
+    })
+    return () => {
+      cancelTracker(correlationId)
+    }
+  }, [correlationId])
```

Comment:
> The usual flow is that the effect is triggered by some data. In this case we have `correlationId`. As long as the trigger remains the same the effect is not be executed. Here the effect is executed once but then it self-triggers internally. My concerns:
> 
>   1. canceling the effect when the component is unmounted - we cancel the timer but if the fetch is in progress then it will set the state  anyway which will cause errors. (btw DataProviderHook has most likely similar problem)
> 

Yep, we have no provisions to handle canceling the fetch if the component is unmounted in this effect or in any data provider.  Given the limited scope of the modals in general, I don't see it as much of a problem.

>   2. fetch error will make it stop working - the user will probably just close the dialog but it's not ideal

That's a case that does make sense to handle.  On fetch error, skip calling the tick and just schedule the next one.

>   3. progress of progress - the refresh rate is fixed but the fetch make take a while  - it would be nice to have some indicator that the work is in progress

Well the progress bar is blue indicating the process in ongoing.  Lame I know.  There isn't really a nice place to add a spinner since it would just be running until the tracker sets state 'complete'.  I can add one if there is a good place to add one.

>   4. tight coupling of "data provider" and "scheduler". IMHO separating both would reduce complexity. Additionally we could use standard provider and just trigger it periodically i.e.
> 
> ```js
> const [trigger, setTrigger] = useState(0)
> const { data, fetchError, fetchInProgress } = useDataProvider({ fetchData, trigger})
> 
> useEffect(() => {
>   scheduleTrackTick(TIME, correlationId, ()=> setTrigger(trigger++))
>   return  () => {
>       cancelTracker(correlationId)
>     }
> }
> ```

There isn't much of a difference in how the component would look between those two forms.  I'm not convinced that the `useEffect` hook isn't sufficient.

> 
> As soon as the progress will be part of the Cluster entity we could also trigger re-resh on the main data provider.

Yeah probably, but still have the annoying blinking wizard in modal issue to deal with.